js/ui/toolbar/ui.toolbar.base.js

```diff
@@ -161,10 +161,77 @@ var ToolbarBase = CollectionWidget.inherit({
             labelPaddings = $label.outerWidth() - $label.width(),
             labelMaxWidth = Math.max(freeSpace - widthBeforeSection - widthAfterSection - labelPaddings, 0);
 
+        if($section.hasClass(TOOLBAR_BEFORE_CLASS)) {
+            this._alignSection(this._$beforeSection, labelMaxWidth);
+            return;
+        }
         $label.css("maxWidth", labelMaxWidth);
     },
 
-    _alignCenterSection: function(beforeRect, afterRect) {
+    _alignSection: function($section, maxWidth) {
+        var $labels = $section.find(TOOLBAR_LABEL_SELECTOR),
+            labels = $labels.toArray(),
+            currentWidth = this._getCurrentLabelsWidth(labels),
+            difference = Math.abs(currentWidth - maxWidth),
+            i,
+            $label,
+            currentLabelWidth,
+            labelMaxWidth;
+
+        if(maxWidth < currentWidth) {
+            labels = labels.reverse();
+            for(i = 0; i < labels.length; i++) {
```

Comment:
Can we refactor it? Function is very long and probably you can extract reusable code parts